runtime-library/CMakeLists.txt

```diff
@@ -0,0 +1,118 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+cmake_minimum_required(VERSION 3.21...3.24)
+
+project(IREE_RUNTIME_LIBRARY)
+cmake_policy(SET CMP0069 NEW)
+
+set(IREE_ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/../../iree" CACHE STRING "Main IREE source")
+
+# Enable LTO if supported.
+option(IREERT_ENABLE_LTO "Enable LTO (link time optimization) if supported" ON)
+include(CheckIPOSupported)
+check_ipo_supported(RESULT _ireert_lto_supported OUTPUT error)
+if(IREERT_ENABLE_LTO)
+  if(_ireert_lto_supported)
+    message(STATUS "Enabling LTO")
+    set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
+  else()
+    message(WARNING "LTO not supported by toolchain bit requested (ignored)")
+  endif()
+endif()
+
+# Optionally enabled shared library build mode. This is only supported for
+# development as it precludes important optimizations.
+option(BUILD_SHARED_LIBS "Build development shared runtime library" OFF)
+
+# When building a shared library, further customize:
+#   - Build with hidden visibility by default.
+#   - Unhide API exported functions.
+#   - Mark symbols for dllexport on Windows by default.
+if(BUILD_SHARED_LIBS)
+  message(STATUS "IREE runtime library with BUILD_SHARED_LIBS=ON is only supported for development!")
+  set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+  add_compile_definitions("IREE_API_ENABLE_VISIBILITY")
+  add_compile_definitions("IREE_API_BUILDING_LIBRARY")
+endif()
+
+# Customize defaults.
+option(IREE_BUILD_COMPILER "Disable compiler for runtime-library build" OFF)
+option(IREE_BUILD_SAMPLES "Disable samples for runtime-library build" OFF)
+
+add_subdirectory("${IREE_ROOT_DIR}" "iree_core" EXCLUDE_FROM_ALL)
+
+#-------------------------------------------------------------------------------
+# Build the ireert library
+#-------------------------------------------------------------------------------
+
+# IREE exposes its transitive objects and deps on special target properties.
+set(_RUNTIME_ROOT_TARGET "iree::runtime::impl")
+set(_RUNTIME_OBJECTS "$<REMOVE_DUPLICATES:$<GENEX_EVAL:$<TARGET_PROPERTY:${_RUNTIME_ROOT_TARGET},INTERFACE_IREE_TRANSITIVE_OBJECTS>>>")
+set(_RUNTIME_LIBS "$<REMOVE_DUPLICATES:$<GENEX_EVAL:$<TARGET_PROPERTY:${_RUNTIME_ROOT_TARGET},INTERFACE_IREE_TRANSITIVE_OBJECT_LIBS>>>")
+ # For debugging, write out objects and deps to files.
```

Comment:
Do we need this in stanadlone example? Given the relative minimal target.

Reply:
I figured I'd leave it to inspire the imagination, and it really does help with debugging in my experience. Will also help people shell-script their way to success by scraping it if their case isn't served by this sample as-is (i.e. encourages a bit of hacking).